In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from scipy import stats

from sqlalchemy import create_engine

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [2]:
pd.__version__

'3.0.3'

In [3]:
#configuration
CONFIG = {
    "high_corr_threshold": 0.75,
    "outlier_iqr_multiplier": 1.5,
    "missing_warning_threshold": 0.10,
    "date_columns": [
        "dose_start",
        "dose_end"
    ]
}


In [4]:
CONFIG

{'high_corr_threshold': 0.75,
 'outlier_iqr_multiplier': 1.5,
 'missing_warning_threshold': 0.1,
 'date_columns': ['dose_start', 'dose_end']}

In [5]:
engine = create_engine(
    
     "postgresql://postgres:postgres123@localhost:5432/malaria_db"
)

In [6]:
#load table
query = "SELECT * FROM malaria_cleaned"

df = pd.read_sql(query, engine)

In [7]:
eda_df = df.copy()

print(eda_df.shape)

(730, 9)


In [8]:
#dataset structure
df.info()
df.dtypes
df.shape



<class 'pandas.DataFrame'>
RangeIndex: 730 entries, 0 to 729
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   participant_id           730 non-null    int64         
 1   dose_num                 730 non-null    int64         
 2   dose_ordered             729 non-null    float64       
 3   dose_start               730 non-null    datetime64[us]
 4   dose_start_time          730 non-null    str           
 5   dose_end                 541 non-null    datetime64[us]
 6   dose_end_time            730 non-null    str           
 7   dose_complete            730 non-null    int64         
 8   treatment_duration_days  541 non-null    float64       
dtypes: datetime64[us](2), float64(2), int64(3), str(2)
memory usage: 51.5 KB


(730, 9)

In [9]:
#detecting duplicate records
duplicate_count = df.duplicated().sum()

print(duplicate_count)

0


 No significant duplicate records were detected in the dataset, indicating good data integrity at the record level.

In [10]:
#checking on missing values
missing_summary = pd.DataFrame({
    "missing_count": df.isnull().sum(),
    "missing_pct": df.isnull().mean()
})

missing_summary.sort_values(
    "missing_pct",
    ascending=False
)

,missing_count,missing_pct
treatment_duration_days,189,0.258904
dose_end,189,0.258904
dose_ordered,1,0.001370
participant_id,0,0.000000
dose_num,0,0.000000
dose_start_time,0,0.000000
dose_start,0,0.000000
dose_end_time,0,0.000000
dose_complete,0,0.000000


The dataset shows a generally strong data quality structure with most fields having complete records.

However, two critical variables — dose_end and treatment_duration_days — show significant missingness at approximately 25.9%.

This missingness is likely structured and not random, as both fields share identical missing counts. This suggests that missing treatment duration is directly linked to missing dose completion timestamps, possibly due to incomplete treatments, ongoing treatments at the time of extraction, or system-level recording gaps.

All other variables, including patient identifiers and dose start information, are fully complete, indicating reliable data capture at the initiation of dosing events.

The minor missing value in dose_ordered is negligible and does not impact overall data integrity.

In [11]:
#isolating missing_dose rows
missing_end = df[df['dose_end'].isna()]
missing_end.head()

,participant_id,dose_num,dose_ordered,dose_start,dose_start_time,dose_end,dose_end_time,dose_complete,treatment_duration_days
5,2,2,220.000000,2019-04-11,TRUE,NaT,FALSE,1,NaN
6,2,3,220.000000,2019-04-12,TRUE,NaT,FALSE,1,NaN
7,2,4,220.000000,2019-04-13,TRUE,NaT,FALSE,1,NaN
8,3,1,66.720001,2019-04-13,TRUE,NaT,FALSE,1,NaN
9,3,2,66.720001,2019-04-14,TRUE,NaT,FALSE,1,NaN


In [12]:
#checking if treatment is ongoing
#checking time patterns
missing_end[['dose_start', 'dose_end_time', 'treatment_duration_days']].head(10)

,dose_start,dose_end_time,treatment_duration_days
5,2019-04-11,FALSE,NaN
6,2019-04-12,FALSE,NaN
7,2019-04-13,FALSE,NaN
8,2019-04-13,FALSE,NaN
9,2019-04-14,FALSE,NaN
10,2019-04-14,FALSE,NaN
11,2019-04-15,FALSE,NaN
64,2019-05-29,FALSE,NaN
65,2019-05-30,FALSE,NaN
66,2019-05-30,FALSE,NaN


In [13]:
#cmparing to time max
df['dose_start'].max()


Timestamp('2019-12-30 00:00:00')

In [14]:
missing_end['dose_start'].describe()

count                           189
mean     2019-08-31 22:20:57.142857
min             2019-04-11 00:00:00
25%             2019-07-23 00:00:00
50%             2019-08-31 00:00:00
75%             2019-10-04 00:00:00
max             2019-12-29 00:00:00
Name: dose_start, dtype: object

In [15]:
#detecting dropout cases
missing_end.groupby('participant_id').size().sort_values(ascending=False).head()

participant_id
3      4
17     4
164    4
29     4
32     4
dtype: int64

In [16]:
#participant history
df[df['participant_id'] == 'some_id'].sort_values('dose_start')

,participant_id,dose_num,dose_ordered,dose_start,dose_start_time,dose_end,dose_end_time,dose_complete,treatment_duration_days


In [17]:
#detecting system errors
missing_end['dose_num'].value_counts()

dose_num
1    48
2    46
3    46
4    45
5     1
6     1
7     1
8     1
Name: count, dtype: int64

In [18]:
#distributin across participants
missing_end['participant_id'].nunique()
df['participant_id'].nunique()

195

many participants are affected suggesting system error

In [19]:
#checking is missingness correlates with time
df.groupby(df['dose_start'].dt.date)['dose_end'].apply(lambda x: x.isna().mean())

dose_start
2019-04-05    0.000000
2019-04-06    0.000000
2019-04-07    0.000000
2019-04-11    0.500000
2019-04-12    1.000000
                ...   
2019-12-25    0.000000
2019-12-26    0.000000
2019-12-28    0.666667
2019-12-29    0.166667
2019-12-30    0.000000
Name: dose_end, Length: 214, dtype: float64

concentratea

In [20]:
#checking relationships between missing fields
df[df['dose_end'].isna()][['dose_end', 'treatment_duration_days']].isna().all()

dose_end                   True
treatment_duration_days    True
dtype: bool

This shows dependency structure

In [21]:
#classification rule creation
def classify_missing(row):
    if pd.isna(row['dose_end']) and pd.isna(row['treatment_duration_days']):
        return "Incomplete record"
    elif pd.isna(row['dose_end']) and not pd.isna(row['dose_start']):
        return "Possible ongoing or dropout"
    else:
        return "Complete"

In [22]:
df['missing_type'] = df.apply(classify_missing, axis=1)
df['missing_type'].value_counts()

missing_type
Complete             541
Incomplete record    189
Name: count, dtype: int64

In [23]:
df.dtypes

participant_id                      int64
dose_num                            int64
dose_ordered                      float64
dose_start                 datetime64[us]
dose_start_time                       str
dose_end                   datetime64[us]
dose_end_time                         str
dose_complete                       int64
treatment_duration_days           float64
missing_type                          str
dtype: object

In [24]:
df['dose_end_time'].unique()

<StringArray>
['TRUE', 'FALSE']
Length: 2, dtype: str

The column dose_end_time was expected to contain time values; however, profiling revealed that it contains only the string values TRUE and FALSE. This indicates that the variable functions as a Boolean status indicator rather than a timestamp field. The column name may therefore be misleading and should be interpreted cautiously during analysis.

In [25]:
df[['dose_end_time', 'dose_end']].head(20)

,dose_end_time,dose_end
0,TRUE,2019-04-05
1,TRUE,2019-04-06
2,TRUE,2019-04-06
3,TRUE,2019-04-07
4,FALSE,2019-04-11
5,FALSE,NaT
6,FALSE,NaT
7,FALSE,NaT
8,FALSE,NaT
9,FALSE,NaT


In [26]:
pd.crosstab( df['dose_end_time'], df['dose_end'].isna())

dose_end,False,True
dose_end_time,,
FALSE,4,188
TRUE,537,1


 Analysis of dose_end_time revealed that the variable functions as a Boolean indicator rather than a timestamp field. In 725 out of 730 records, the variable correctly aligns with the presence or absence of a dose_end value. However, five records exhibit inconsistencies where the indicator and end-date information conflict. These records should be flagged for further review prior to downstream analyses involving treatment completion or duration calculations.

In [27]:
#investigating the 5 inconsistent records
df[
    ((df['dose_end_time'] == 'FALSE') & (df['dose_end'].notna())) |
    ((df['dose_end_time'] == 'TRUE') & (df['dose_end'].isna()))
]

,participant_id,dose_num,dose_ordered,dose_start,dose_start_time,dose_end,dose_end_time,dose_complete,treatment_duration_days,missing_type
4,2,1,NaN,2019-04-11,TRUE,2019-04-11,FALSE,1,0.0,Complete
14,4,3,163.0,2019-04-16,TRUE,2019-04-16,FALSE,1,0.0,Complete
176,45,3,120.0,2019-07-04,TRUE,2019-07-04,FALSE,1,0.0,Complete
441,113,1,155.0,2019-09-03,FALSE,2019-09-03,FALSE,1,0.0,Complete
667,176,1,155.0,2019-11-25,TRUE,NaT,TRUE,1,NaN,Incomplete record


Data Quality Finding: Completion Flag Consistency
Cross-validation between dose_end_time and dose_end revealed five inconsistent records (0.68% of the dataset). Four records were marked as FALSE in dose_end_time despite containing valid dose end dates, while one record was marked as TRUE despite having a missing end date. Given that these inconsistencies represent less than 1% of all records, the variable generally behaves as a completion indicator. However, the identified records should be flagged as data quality exceptions and reviewed before conducting adherence or treatment-duration analyses.


In [28]:
#Date Validatin
df["dose_start"] = pd.to_datetime(
    df["dose_start"]
)

df["dose_end"] = pd.to_datetime(
    df["dose_end"]
)

In [29]:
#verifying conversin of dates
df[["dose_start","dose_end"]].dtypes

dose_start    datetime64[us]
dose_end      datetime64[us]
dtype: object

Both dose_start and dose_end were successfully converted to datetime format (datetime64[us]). This confirms that the variables are suitable for temporal analysis, duration calculations, chronological sequencing, and date-based validation checks.

In [30]:
#checking logical incnsistencies
invalid_dates = df[df['dose_end'] < df['dose_start']]

invalid_dates

,participant_id,dose_num,dose_ordered,dose_start,dose_start_time,dose_end,dose_end_time,dose_complete,treatment_duration_days,missing_type


In [31]:
(df['dose_end'] < df['dose_start']).sum()

np.int64(0)

In [32]:
#verifying duration incnsistency
(
    df['dose_end'] - df['dose_start']
).describe()

count                       541
mean     0 days 01:11:52.014787
std      0 days 20:51:09.222029
min             0 days 00:00:00
25%             0 days 00:00:00
50%             0 days 00:00:00
75%             0 days 00:00:00
max            20 days 00:00:00
dtype: object

In [33]:
duration = df['dose_end'] - df['dose_start']

df[duration > pd.Timedelta(days=1)]

,participant_id,dose_num,dose_ordered,dose_start,dose_start_time,dose_end,dose_end_time,dose_complete,treatment_duration_days,missing_type
241,63,4,150.000000,2019-07-27,TRUE,2019-07-29,TRUE,1,2.0,Complete
258,68,1,123.800003,2019-07-08,TRUE,2019-07-28,TRUE,1,20.0,Complete


Duration validation identified only two records with durations greater than one day. One record exhibited a duration of 2 days, while another showed a duration of 20 days. Given that the majority of records have same-day start and end dates, the 20-day duration represents a notable outlier and warrants further investigation. No negative durations were identified, indicating that dose end dates generally occur on or after dose start dates.

In [34]:
#investigating participant_id 68
df[df['participant_id'] == 68]


,participant_id,dose_num,dose_ordered,dose_start,dose_start_time,dose_end,dose_end_time,dose_complete,treatment_duration_days,missing_type
258,68,1,123.800003,2019-07-08,TRUE,2019-07-28,TRUE,1,20.0,Complete
259,68,2,123.800003,2019-07-29,TRUE,2019-07-29,TRUE,1,0.0,Complete
260,68,3,123.800003,2019-07-29,TRUE,2019-07-29,TRUE,1,0.0,Complete
261,68,4,123.800003,2019-07-30,TRUE,2019-07-30,TRUE,1,0.0,Complete


Extended duration review
    Review of records with durations greater than one day identified two cases. One record showed a duration of two days, while participant 68 exhibited a duration of twenty days for the first dose. The calculated duration is internally consistent with the recorded start and end dates and matches the stored treatment duration value. Although the record represents an outlier relative to the rest of the dataset, there is insufficient evidence to classify it as erroneous. The observation should therefore be flagged as an unusual but valid record pending domain-specific review.

In [35]:
# Check for dates before start dates
(df['dose_end'] < df['dose_start']).sum()


np.int64(0)

In [36]:
# Check overall date ranges
df[['dose_start', 'dose_end']].agg(['min', 'max'])

,dose_start,dose_end
min,2019-04-05,2019-04-05
max,2019-12-30,2019-12-30


Date range validation
 The date variables span from 5 April 2019 to 30 December 2019. Both dose_start and dose_end share the same minimum and maximum dates, indicating consistency in the recorded treatment timeline. No implausible dates were identified within the observation period.

Data validation findings:
    dose_start and dose_end were successfully converted to datetime format.
Missing end dates account for approximately 25.9% of records.
No evidence of widespread date-format issues.
Most dosing events start and end on the same day.
Only two records have durations greater than one day (2 days and 20 days).
Date ranges are consistent and fall between April and December 2019.
No implausible calendar dates were identified.
A small number of records show inconsistencies between dose_end_time and dose_end.

Interpretation
    Overall, the date variables demonstrate good quality and logical consistency. The primary concern is missing end-date information rather than invalid dates. Apart from a few documented anomalies and extended-duration outliers, the temporal structure of the dataset appears suitable for downstream analysis.

In [37]:
#detecting impossible reccords
invalid_duration = df[
    df["treatment_duration_days"] < 0
]

invalid_dates = df[
    df["dose_end"] < df["dose_start"]
]

invalid_dose_number = df[
    df["dose_num"] <= 0
]

In [38]:
#verification
len(invalid_duration)
len(invalid_dates)
len(invalid_dose_number)

0

Data Quality Validation Results

• No records were found with negative treatment durations.
• No records were found where dose_end occurred before dose_start.
• No records were found with invalid dose numbers.

Interpretation:
The dataset passed all predefined logical validation checks. Temporal relationships between dosing events are consistent, treatment durations are valid, and dose numbering follows expected clinical rules.

Conclusion:
The dataset demonstrates a high level of structural and logical integrity and is suitable for exploratory data analysis.

In [39]:
#Numeric summary
numeric_summary = df.describe().T
numeric_summary

,count,mean,min,25%,50%,75%,max,std
participant_id,730.0,95.223288,1.0,48.0,94.0,142.0,197.0,55.65572
dose_num,730.0,2.423288,1.0,1.0,2.0,3.0,8.0,1.133239
dose_ordered,729.0,163.77347,23.5,120.0,175.0,210.0,360.0,68.229778
dose_start,730,2019-08-21 12:43:23.835616,2019-04-05 00:00:00,2019-07-06 06:00:00,2019-08-18 00:00:00,2019-09-30 18:00:00,2019-12-30 00:00:00,NaN
dose_end,541,2019-08-17 22:42:48.576709,2019-04-05 00:00:00,2019-06-30 00:00:00,2019-08-13 00:00:00,2019-09-22 00:00:00,2019-12-30 00:00:00,NaN
dose_complete,730.0,1.0,1.0,1.0,1.0,1.0,1.0,0.0
treatment_duration_days,541.0,0.049908,0.0,0.0,0.0,0.0,20.0,0.868857



## Insight on participant_id

1. Participant IDs range from 1 to 197, indicating the dataset contains records from up to 197 unique participants.
2. The mean participant ID (95.2) has no analytical meaning because participant IDs are identifiers rather than measured variables.
3. This variable should be used for grouping and participant-level analysis rather than statistical interpretation.
Conclusion
Participant IDs appear to be sequential and within a reasonable range.

## Insight on dose_numbers

1. Participants received between 1 and 8 recorded doses.
2. The median dose number of 2 indicates that most observations correspond to early treatment doses.
3. The relatively low standard deviation (1.13) suggests limited variability in dose sequence positions.
4. The upper quartile of 3 indicates that 75% of recorded doses occur within the first three administrations.
Conclusion
The dataset is dominated by early treatment doses, with relatively few observations from later doses.

## Insight on dose_ordered

1. One record is missing a dose ordering value (729 vs. 730 records).
2. Ordered doses range from 23.5 to 360, indicating substantial variation in prescribed dosing quantities.
3. The mean and median are relatively close, suggesting a fairly balanced distribution.
4. The large standard deviation (68.2) indicates considerable variability across treatment prescriptions.
Conclusion
Prescribed dose quantities vary substantially across treatment events and may warrant further investigation by participant or treatment period.

## Insight on dose

1. Treatment records span approximately April to December 2019.
2. Half of all treatment initiations occurred before mid-August 2019 and half occurred afterward.
3. The dataset provides coverage across multiple months, making temporal trend analysis feasible.
Conclusion
The data contains sufficient temporal coverage for longitudinal and seasonal analyses.

## Insight on dose_end_dates

1. Only 541 of 730 records contain dose end dates.
2. This means 189 records (25.9%) have missing completion timestamps.

Calculation:

(730 - 541) / 730 × 100
= 25.9%
Conclusion
Missing end dates represent a significant data quality issue that should be considered when analyzing treatment completion and duration metrics.

## Insight on date_competion

1. Every record has a completion value of 1.
2. There is no variation in this variable.
3. The variable provides no discriminatory analytical value because all observations share the same value.
Conclusion
dose_complete can be excluded from most analyses since it contains no variability.

## Treatment duration insight:

1. Treatment duration is highly concentrated at zero days.
2. Since the median and 75th percentile are both zero, at least 75% of completed doses occurred within the same calendar day.
3. The maximum duration of 20 days is substantially larger than the majority of observations.
4. The mean (0.05 days ≈ 1.2 hours) is much higher than the median due to a small number of long-duration records.
4. This pattern suggests a highly right-skewed distribution.
Conclusion
Most treatment administrations are completed on the same day, but a small number of unusually long-duration events may represent outliers or special clinical circumstances requiring further investigation.

The dataset contains 730 medication administration records spanning April–December 2019. Most treatment events occur within the first three dose administrations and are completed within the same day. While dose sequencing appears logically consistent, approximately 26% of records lack dose end dates, which may affect duration-based analyses. Treatment duration is highly right-skewed, with a small number of unusually long-duration events potentially representing outliers. Overall, the dataset appears suitable for exploratory analysis, although missing completion timestamps should be addressed before conducting duration-related investigations.





# exporting profiled dataset

In [40]:
eda_df.to_sql(
    "malaria_profiled",
    engine,
    if_exists="replace",
    index=False
)

730